wgsl/index.bs

```diff
@@ -4557,11 +4555,271 @@ TODO: *Stub*: Expression evaluation
+Each function is analyzed in two phases.
+
+The first phase walks over the AST of the function, building a directed graph along the way.
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+Note: Each vertex in this graph corresponds either to the statement "The control-flow at a specific program point must be uniform" or to "The value of a specific expression must be uniform" or to "A specific variable must always hold a uniform value" or is one of the special vertices True and False.
+      Each (directed) edge corresponds to an implication.
+
+In more detail, here is the algorithm for analyzing a function:
+
+* Create some vertices called "True", "False", "CF_start", "CF_return" and if the function is non-void a vertex called "Value_return"
+* Create one vertex for each parameter of the function which we'll call "arg_i"
+* Walk over the AST of the function, adding vertices and edges to the graph following the rules of the next section.
+* Look at the *Behaviors* of the function body treated as a statement (see next section):
+    * If it has a behavior which is neither Nothing nor Discard, then reject the program
+    * Else remember it for later it.
+* Look at which vertices are reachable from "True"
+    * If this set includes the vertex "False", then reject the program.
+    * Remember for later whether this set includes "CF_start" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "CF_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "Value_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. Also note that the information that we need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Nothing}.
+
+TODO: do we need an extra behavior for possibly infinite loops? I should check whether some form of reconvergence happens after something like `while(tid mod 2) {}`.
+
+If the set of behaviors for a statement is a singleton, we take that to mean that reconvergence happened, and we pick the vertex corresponding to control-flow at the start of the statement as the vertex corresponding to control-flow at the exit of the statement, overriding the other rules below.
```

Comment:
I think the use of the word "singleton" here is a bit confusing for programmers, because of the design pattern of the same name. It might be worth rephrasing this to be "If the size of the set of behaviors for a statement is 1, ..." (Please ignore this comment if "singleton" is a term-of-art in set theory.)

Reply:
I think that "singleton" is a term of art, but being clear to programmers who read the spec is more important! I'll modify this.